### ISSUE TITLE
Achieve 20% test coverage

### ISSUE THREAD

**Issue by carlschroedl** [happyjack27/autoredistrict#40]
We need to ensure that the app works and continues to work as it grows. Insert all normal reasons for testing here. We could do 20% in terms of line coverage, or branch coverage, either is good with me; this is just a milestone on a march to a more stable codebase. Should probably use JUnit to run the tesets. I'm not sold on any particular tool to measure coverage. I've heard coveralls is cool.

**Comment by happyjack27** [comment]
i'd like to focus on enhancements and bug fixes now. (not so concerned
about #32 though, that can wait.)  i made a "To do - priority" category.

**Comment by carlschroedl** [comment]
As a prospective contributor, I would like to work on enhancements and bug fixes too, but I cannot do so quickly and confidently without reading large swaths of the codebase and performing tiresome error-prone manual tests. Deferring implementation of tests defers the ability of others to effectively contribute to this project. If not now, then when should we implement this very modest test coverage goal?

**Comment by carlschroedl** [comment]
I suspect that prospective users (legislature, judicial committee, citizen committee, etc.) will be reluctant to adopt this software without positively awesome test coverage.

**Comment by happyjack27** [comment]
i assure you their reluctance will be for reasons not nearly as
sophisticated.

**Comment by happyjack27** [comment]
test coverage is something that seems to me will take a long time.

automatic tests are more error-prone than manual tests, since they only
cover a small subset of what can go wrong. all changes must be manually
tested.   every individual feature addition or what-not should be manually
tested immediately.   automatic tests are no substitute for this.

Deferring implementation of tests does NOT defer the ability of others to
effectively contribute to this project.

i will be happy to guide you through the code.  (one thing that would be
nice is a sort of map of the code.)

i'd like to get the priority items out of the way, as this software is
being used now and it is a waste of users time to not get those finished.
 these will also help you get comfortable with the code.  pick one and i
can guide you through and answer any questions.

**Comment by carlschroedl** [comment]
This is how I felt when I read "automatic tests are more error-prone than manual tests":
![wow](https://i.imgur.com/KBNcZh.jpg)

What is so spectacularly special about this project that makes it immune from mainstream software best practices?

**Comment by carlschroedl** [comment]
Sorry, I'm frustrated about our differences of opinion. Lest this spiral out of control, I want to I thank you for your offer to guide me through the code. I re-iterate my thanks to you for working on this project in the first place.

**Comment by happyjack27** [comment]
nothing at all.

there's nothing special about "software best practices", either.  just
because they are "best practices" doesn't mean they're good ideas or that
they are always top priority.

appeal to population holds no sway.  http://skepdic.com/adpopulum.html

decisions are to be made by a rational and specific consideration of
specific merits, not whether or not they are popular.

**Comment by happyjack27** [comment]
if you want to work on the test case thing, go right ahead.  i'm just
saying that I have other priorities right now, and I'm happy to help you
get your feet wet.

**Comment by carlschroedl** [comment]
Ok. Thanks for clarifying. I want to be clear that I'm not advocating for increasing test coverage because it's popular. I'm drawing on experience from dozens of enterprise projects in which testing has been a worthwhile investment. The automated tests ensured that the software was correct and continued to be correct as features were added, removed and refactored. I view the correctness of this software as equally or more important than an enterprise app, so I want to make a dent in test coverage to increase assurance that the software functions correctly, and continues to do so. I'm glad that you would welcome such contributions. I understand that you are not interested in accomplishing that task right now.

**Comment by happyjack27** [comment]
yeah, i think that's accurate.  and i do recognize the value of automated
tests and i like it, and i welcome that and appreciate it.

**Comment by carlschroedl** [comment]
Awesome! Thanks! Also, I'm glad to hear that we have some communication with a user group. I'd like to learn more about them!

**Comment by happyjack27** [comment]
Well, sounds like 1 at the moment, but it's a big one.  This site:
http://www.fairvote.org/  and this organization:
https://en.wikipedia.org/wiki/FairVote They are a nation-wide organization
focused on advocating for multi-member districts.  (I think they're focus
is a little broader, but that's their main focus right now, i believe.)
 Staff of 11, according to wikipedia.

their senior policy analyst got in touch with me a few months ago.  they
want to use the software to design districts for all 50 states, and they're
planning on doing a write up on it, at the bare minimum in their blog:
http://www.fairvote.org/blog_hub    So it's potentially huge press.

they like the program for it's focus on fairness.  when they first
contacted me, the program could do multi-member, but not the way they
wanted, they wanted some 5 seat districts and some 3 seat districts in the
same map.  so i added that in for them along with a few other tweaks.
they've since provided a few more suggestions which you can see on the
board.  i think it's an awesome relationship - very mutually beneficial.
 it occurs to me that redistricting all 50 states probably would be
impossible for an organization that size without this software.  state
governments often spend millions to redistrict just one state.  (where all
that money goes, your guess is as good as mine.  suffice it to say if
they're bidding it out i'd be happy to be the lowball bidder.)  anyways,
any time we can save fairvote, multiply that by 50.

**Comment by Noreaster76** [comment]
I assume this issue still applies, correct?

**Comment by happyjack27** [comment]
I don't even know what it means.  You're welcome to work towards it.
